flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/program/FlinkChainedPrograms.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.optimize.program
+
+import org.apache.flink.table.api.TableException
+import org.apache.flink.table.util.Logging
+import org.apache.flink.util.Preconditions
+
+import org.apache.calcite.plan.RelOptUtil
+import org.apache.calcite.rel.RelNode
+
+import java.util
+
+import scala.collection.JavaConversions._
+
+
+/**
+  * FlinkChainedPrograms contains a sequence of [[FlinkOptimizeProgram]]s which are chained
+  * together.
+  *
+  * The chained-order of programs can be adjusted by [[addFirst]], [[addLast]], [[addBefore]]
+  * and [[remove]] methods.
+  *
+  * When [[optimize]] method called, each program's optimize method will be called in sequence.
+  *
+  * @tparam OC OptimizeContext
+  */
+class FlinkChainedPrograms[OC <: OptimizeContext] extends Logging {
```

Comment:
This class should also extends FlinkOptimizeProgram?

Reply:
That's a good idea!